modules/sagemaker/sagemaker-model-cicd/lib/code-pipeline/model-build/model-build-code-pipeline-stack.ts

```diff
@@ -0,0 +1,152 @@
+  readonly sagemakerExecutionRoleName: string;
+  readonly codeBuildAssumeRoleName: string;
+  readonly modelPackageGroupName: string;
+}
+
+export class ModelBuildCodePipelineStack extends cdk.Stack {
+  public readonly pipeline: codepipeline.Pipeline;
+
+  constructor(
+    scope: Construct,
+    id: string,
+    props: ModelBuildCodePipelineProps,
+  ) {
+    super(scope, id, props);
+
+    const {
+      projectName,
+      sourceBranch,
+      modelBuildEnvironment,
+      modelBuildSourceRepoName,
+      sagemakerArtifactsBucketName,
+      sagemakerExecutionRoleName,
+      codeBuildAssumeRoleName,
+      modelPackageGroupName,
+      pipelineName = `${projectName}-${sourceBranch}-build`,
+      pipelineDescription = `${projectName} model build pipeline for ${sourceBranch} branch`,
+    } = props;
+
+    const codeBuildRole = new iam.Role(this, 'CodeBuildRole', {
+      assumedBy: new iam.ServicePrincipal('codebuild.amazonaws.com'),
+      path: '/service-role/',
+    });
+
+    // CodeBuild will assume this cross-account role in the model build account to start the SageMaker pipeline
+    const codeBuildAssumeRole = iam.Role.fromRoleArn(
+      this,
+      'CodeBuildAssumeRole',
+      `arn:aws:iam::${modelBuildEnvironment.account}:role/${codeBuildAssumeRoleName}`,
+    );
+
+    // this role will be used by model build code (i.e., sagemaker service) to access aws resources in the model build environment
+    const sageMakerExecutionRoleArn = `arn:aws:iam::${modelBuildEnvironment.account}:role/service-role/${sagemakerExecutionRoleName}`;
+
+    const codeBuildPolicy = new iam.Policy(this, 'CodeBuildPolicy', {
+      document: new iam.PolicyDocument({
+        statements: [
+          new iam.PolicyStatement({
+            sid: 'AssumeSageMakerAccessRole',
+            actions: ['sts:AssumeRole'],
+            resources: [codeBuildAssumeRole.roleArn],
+          }),
+          new iam.PolicyStatement({
+            sid: 'LogsAccess',
+            actions: [
+              'logs:CreateLogGroup',
+              'logs:CreateLogStream',
+              'logs:PutLogEvents',
+            ],
+            resources: ['*'],
+          }),
+          new iam.PolicyStatement({
+            sid: 'MetricsAccess',
+            actions: ['cloudwatch:PutMetricData'],
+            resources: ['*'],
+          }),
+        ],
+      }),
+    });
+    codeBuildPolicy.attachToRole(codeBuildRole);
+
+    const smPipelineBuild = new codebuild.PipelineProject(
+      this,
+      'SMPipelineBuild',
+      {
+        projectName: pipelineName,
+        role: codeBuildRole,
+        buildSpec: codebuild.BuildSpec.fromSourceFilename('buildspec.yml'),
+        environment: {
+          buildImage: codebuild.LinuxBuildImage.STANDARD_7_0,
+          environmentVariables: {
+            // set aws sdk calls to correct region incase of cross region pipeline
+            AWS_REGION: { value: modelBuildEnvironment.region },
```

Comment:
Q - is this compatible with existing SM templates, for example [abalone model build](https://github.com/awslabs/aiops-modules/blob/main/modules/sagemaker/sagemaker-templates-service-catalog/templates/xgboost_abalone/pipeline_constructs/build_pipeline_construct.py#L229)?

Reply:
For most part yes. 

**Differences:**
- Some parameters like `SAGEMAKER_PROJECT_ID` and `SAGEMAKER_PROJECT_NAME` are not valid as those are created by sagemaker studio I believe. Although, these [parameters seems to be only used for tagging in this template](https://github.com/awslabs/aiops-modules/blob/main/modules/sagemaker/sagemaker-templates-service-catalog/templates/xgboost_abalone/seed_code/build_app/buildspec.yml#L17), so we could add some default values to these parameters if needed.
- The buildspec.yml in the template needs to be updated to assume the cross account role before starting the pipeline execution
```
# assume the cross account role to allow code build start sagemaker execution
- >
- assumed_role_creds=$(aws sts assume-role
    --role-arn $ASSUME_ROLE_ARN
    --role-session-name "ci-sagemaker-pipeline-execution"
    --output json)
- if [ $? -ne 0 ]; then echo "Failed to assume the role $ASSUME_ROLE_ARN"; exit 1; fi
- export AWS_ACCESS_KEY_ID=$(echo "${assumed_role_creds}" | jq -r .Credentials.AccessKeyId)
- export AWS_SECRET_ACCESS_KEY=$(echo "${assumed_role_creds}" | jq -r .Credentials.SecretAccessKey)
- export AWS_SESSION_TOKEN=$(echo "${assumed_role_creds}" | jq -r .Credentials.SessionToken)
# now start the execution
- run-pipeline ...
```